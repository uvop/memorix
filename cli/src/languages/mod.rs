mod python;
mod rust;
mod typescript;

use crate::{
    flat_schema::FlatValidatedSchema,
    parser::{Export, Language},
    validate::ValidatedSchema,
};
use std::collections::HashMap;

pub fn codegen_per_language(
    export: &Export,
    validated_schema: &ValidatedSchema,
    flat_validated_schema: &FlatValidatedSchema,
) -> HashMap<String, String> {
    export
        .files
        .as_ref()
        .unwrap_or(&vec![])
        .iter()
        .map(|file| {
            let path = file.path.to_string();
            let content = match file.language {
                Language::TypeScript => typescript::codegen(validated_schema),
                Language::Python => python::codegen(flat_validated_schema),
                Language::Rust => rust::codegen(flat_validated_schema),
            };
            (path, content)
        })
        .collect()
}
