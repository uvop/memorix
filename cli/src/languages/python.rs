use crate::{
    flat_schema::{
        FlatValidatedNamespace, FlatValidatedReferenceTypeItemKind, FlatValidatedSchema,
        FlatValidatedTypeItem, TypeObjectItem,
    },
    parser::{Engine, Value, ALL_CACHE_OPERATIONS, ALL_PUBSUB_OPERATIONS, ALL_TASK_OPERATIONS},
};

fn indent(level: usize) -> String {
    "    ".repeat(level)
}
fn flat_type_item_to_code(
    flat_type_item: &FlatValidatedTypeItem,
    schema: &FlatValidatedSchema,
) -> String {
    match flat_type_item {
        FlatValidatedTypeItem::Optional(x) => {
            format!("typing.Optional[{}]", flat_type_item_to_code(x, schema)).to_string()
        }
        FlatValidatedTypeItem::Array(x) => {
            format!("typing.List[{}]", flat_type_item_to_code(x, schema)).to_string()
        }
        FlatValidatedTypeItem::Reference(x) => {
            let (namespace, namespace_names) = x.namespace_indexes.iter().fold(
                (&schema.global_namespace, Vec::<&str>::new()),
                |(current_namespace, mut acc), &i| {
                    let (next_name, next_namespace) = &current_namespace.namespaces[i];
                    acc.push(next_name.as_str());
                    (next_namespace, acc)
                },
            );
            format!(
                "{prefix}{seperator}{name}",
                prefix = namespace_names
                    .iter()
                    .map(|x| stringcase::pascal_case(x))
                    .collect::<Vec<_>>()
                    .join("."),
                seperator = match namespace_names.is_empty() {
                    true => "",
                    false => ".",
                },
                name = match x.kind {
                    FlatValidatedReferenceTypeItemKind::TypeItem(i) => {
                        namespace.flat_type_items[i].0.clone()
                    }
                    FlatValidatedReferenceTypeItemKind::TypeObjectItem(i) => {
                        namespace.type_object_items[i].0.clone()
                    }
                    FlatValidatedReferenceTypeItemKind::Enum(i) => {
                        namespace.enum_items[i].0.clone()
                    }
                }
            )
        }
        FlatValidatedTypeItem::Boolean => "bool".to_string(),
        FlatValidatedTypeItem::String => "str".to_string(),
        FlatValidatedTypeItem::U32 => "int".to_string(),
        FlatValidatedTypeItem::U64 => "int".to_string(),
        FlatValidatedTypeItem::I32 => "int".to_string(),
        FlatValidatedTypeItem::I64 => "int".to_string(),
        FlatValidatedTypeItem::F32 => "float".to_string(),
        FlatValidatedTypeItem::F64 => "float".to_string(),
    }
}

fn value_to_code(value: &Value) -> String {
    match value {
        Value::String(x) => format!("\"{x}\""),
        Value::Env(x) => {
            format!("os.environ[\"{x}\"]")
        }
    }
}

fn type_item_object_to_code(
    name: &str,
    type_item_object: &TypeObjectItem,
    schema: &FlatValidatedSchema,
    level: usize,
) -> String {
    let base_indent = indent(level);
    format!(
        r#"
{base_indent}@dataclass
{base_indent}class {name}(object):
{}
"#,
        type_item_object
            .properties
            .iter()
            .map(|(property_name, flat_type_item)| format!(
                "{base_indent}    {property_name}: {}",
                flat_type_item_to_code(flat_type_item, schema)
            )
            .to_string())
            .collect::<Vec<_>>()
            .join("\n")
    )
    .to_string()
}

fn namespace_to_code(
    namespace: &FlatValidatedNamespace,
    name_tree: Vec<String>,
    schema: &FlatValidatedSchema,
) -> String {
    let mut result = String::from("");
    let base_indent = indent(name_tree.len());

    for (name, values) in &namespace.enum_items {
        result.push_str(&format!(
            r#"
{base_indent}class {name}(str, Enum):
{}

"#,
            values
                .iter()
                .map(|x| format!("{base_indent}    {x} = \"{x}\""))
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }
    for (name, type_item_object) in &namespace.type_object_items {
        result.push_str(&format!(
            "{}\n",
            type_item_object_to_code(name.as_str(), type_item_object, schema, name_tree.len())
        ));
    }
    for (name, flat_type_item) in &namespace.flat_type_items {
        result.push_str(&format!(
            "{base_indent}{name} = {}\n\n",
            flat_type_item_to_code(flat_type_item, schema)
        ));
    }
    for (name, sub_namespace) in &namespace.namespaces {
        result.push_str(&format!(
            r#"{base_indent}class {name}(object):
{namespace_content}
"#,
            name = stringcase::pascal_case(name),
            namespace_content = namespace_to_code(
                sub_namespace,
                name_tree
                    .clone()
                    .into_iter()
                    .chain(std::iter::once(name.clone()))
                    .collect(),
                schema,
            ),
        ));
    }
    if !namespace.cache_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixCache(MemorixCacheAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .cache_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixCacheAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},{options}
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_CACHE_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join(""),
                        options = {
                            let content = [
                                item.ttl.as_ref().map(|x| {
                                    format!(
                                        "{base_indent}                ttl={},",
                                        value_to_code(x)
                                    )
                                }),
                                item.extend_on_get.as_ref().map(|x| {
                                    format!(
                                        "{base_indent}                extend_on_get={},",
                                        value_to_code(x)
                                    )
                                }),
                            ]
                            .into_iter()
                            .flatten()
                            .collect::<Vec<_>>()
                            .join("\n");
                            match content.is_empty() {
                                true => "".to_string(),
                                false => format!(
                                    r#"
{base_indent}            options=MemorixCacheAll.Options(
{content}
{base_indent}            )"#
                                ),
                            }
                        }
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    if !namespace.pubsub_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixPubSub(MemorixPubSubAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .pubsub_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixPubSubAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_PUBSUB_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join("")
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    if !namespace.task_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixTask(MemorixTaskAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .task_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixTaskAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},{options}
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_TASK_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join(""),
                        options = {
                            let content = [item.queue_type.as_ref().map(|x| {
                                format!(
                                    "{base_indent}                queue_type={},",
                                    value_to_code(x)
                                )
                            })]
                            .into_iter()
                            .flatten()
                            .collect::<Vec<_>>()
                            .join("\n");
                            match content.is_empty() {
                                true => "".to_string(),
                                false => format!(
                                    r#"
{base_indent}            options=MemorixTaskAll.Options(
{content}
{base_indent}            )"#
                                ),
                            }
                        }
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    let pre_namespace = match name_tree.is_empty() {
        true => "".to_string(),
        false => format!(
            "{}.",
            name_tree
                .iter()
                .map(|x| stringcase::pascal_case(x))
                .collect::<Vec<_>>()
                .join(".")
        ),
    };
    result.push_str(&format!(
        r#"{base_indent}class Memorix(MemorixBase):
{init_def}

{base_indent}        self._namespace_name_tree = [{name_tree}]
{class_content}
"#,
        name_tree = name_tree
            .iter()
            .map(|x| format!("\"{}\"", x))
            .collect::<Vec<_>>()
            .join(", "),
        init_def = match name_tree.is_empty() {
            true => format!(
                r#"{base_indent}    def __init__(self) -> None:
{base_indent}        super().__init__(redis_url={redis_url})
"#,
                redis_url = match &schema.engine {
                    Engine::Redis(x) => value_to_code(x),
                }
            ),
            false => format!(
                r#"{base_indent}    def __init__(self, ref: MemorixBase) -> None:
{base_indent}        super().__init__(ref=ref)
"#
            ),
        },
        class_content = [
            namespace
                .namespaces
                .iter()
                .map(|(name, _)| format!(
                    "{base_indent}        self.{name} = {pre_namespace}{namespace_name}.Memorix(ref=self)",
                    namespace_name = stringcase::pascal_case(name)
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            [
                (!namespace.cache_items.is_empty())
                    .then(|| format!("{base_indent}        self.cache = {pre_namespace}MemorixCache(self)")),
                (!namespace.pubsub_items.is_empty())
                    .then(|| format!("{base_indent}        self.pubsub = {pre_namespace}MemorixPubSub(self)")),
                (!namespace.task_items.is_empty())
                    .then(|| format!("{base_indent}        self.task = {pre_namespace}MemorixTask(self)")),
            ]
            .into_iter()
            .flatten()
            .collect::<Vec<_>>()
            .join("\n")
        ]
        .into_iter()
        .filter(|x| !x.is_empty())
        .collect::<Vec<_>>()
        .join("\n\n"),
    ));
    result
}

pub fn codegen(schema: &FlatValidatedSchema) -> String {
    format!(
        r#"# flake8: noqa
import typing
import os

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from enum import Enum
from memorix_client_redis import (
    MemorixBase,
    MemorixCacheAll,
    MemorixPubSubAll,
    MemorixTaskAll,
)

{}"#,
        namespace_to_code(&schema.global_namespace, vec![], schema)
    )
    .to_string()
}
