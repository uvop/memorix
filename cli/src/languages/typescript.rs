use crate::{
    export_schema::{ExportNamespace, ExportSchema},
    parser::TypeItem,
    parser::{Engine, Value, ALL_CACHE_OPERATIONS, ALL_PUBSUB_OPERATIONS, ALL_TASK_OPERATIONS},
};

fn indent(level: usize) -> String {
    "  ".repeat(level)
}
fn type_item_to_code(type_item: &TypeItem, level: usize) -> String {
    let base_indent = indent(level);

    match type_item {
        TypeItem::Object(x) => format!(
            r#"{{
{}
{base_indent}}}"#,
            x.iter()
                .map(|(property_name, type_item)| format!(
                    "{base_indent}  {property_name}: {};",
                    type_item_to_code(type_item, level + 1)
                )
                .to_string())
                .collect::<Vec<_>>()
                .join("\n")
        ),
        TypeItem::Optional(x) => format!("undefined | {}", type_item_to_code(x, level)).to_string(),
        TypeItem::Array(x) => format!(
            r#"Array<
{}
{base_indent}>"#,
            type_item_to_code(x, level + 1)
        )
        .to_string(),
        TypeItem::Reference(x) => x.clone(),
        TypeItem::Boolean => "boolean".to_string(),
        TypeItem::String => "string".to_string(),
        TypeItem::U32 => "number".to_string(),
        TypeItem::U64 => "number".to_string(),
        TypeItem::I32 => "number".to_string(),
        TypeItem::I64 => "number".to_string(),
        TypeItem::F32 => "number".to_string(),
        TypeItem::F64 => "number".to_string(),
    }
}

fn value_to_code(value: &Value) -> String {
    match value {
        Value::String(x) => format!("\"{x}\""),
        Value::Env(x) => {
            format!("getEnvVariable(\"{x}\")")
        }
    }
}

fn namespace_to_code(
    namespace: &ExportNamespace,
    name_tree: Vec<String>,
    engine: &Engine,
) -> String {
    let mut result = String::from("");
    let level = name_tree.len();
    let base_indent = indent(level);

    for (name, values) in &namespace.enum_items {
        result.push_str(&format!(
            r#"{base_indent}export enum {name} {{
{}
}}

"#,
            values
                .iter()
                .map(|x| format!("{base_indent}  {x} = \"{x}\","))
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }
    for (name, type_item) in &namespace.type_items {
        result.push_str(&format!(
            "{base_indent}export type {name} = {};\n\n",
            type_item_to_code(type_item, level)
        ));
    }
    for (name, sub_namespace) in &namespace.namespaces {
        result.push_str(&format!(
            r#"{base_indent}export namespace {name} {{
{namespace_content}
{base_indent}}}
"#,
            name = stringcase::camel_case(name),
            namespace_content = namespace_to_code(
                sub_namespace,
                name_tree
                    .clone()
                    .into_iter()
                    .chain(std::iter::once(name.clone()))
                    .collect(),
                engine,
            ),
        ));
    }
    result.push_str(&format!(
        r#"{base_indent}export class Memorix extends MemorixBase {{{start}
{base_indent}  protected override namespaceNameTree = [{name_tree}];

{class_content}
{base_indent}}}"#,
start= match name_tree.is_empty() {
      true => format!(r#"
{base_indent}  protected override redisUrl = {redis_url};
"#, redis_url = match engine {
    Engine::Redis(x) => value_to_code(x),
    }),
    false => "".to_string(),
},
        name_tree = name_tree
            .iter()
            .map(|x| format!("\"{}\"", x))
            .collect::<Vec<_>>()
            .join(", "),
        class_content = [
            namespace
                .namespaces
                .iter()
                .map(|(name, _)| format!(
                    "{base_indent}  {name} = this.getNamespaceItem({namespace_name}.Memorix);",
                    namespace_name = stringcase::camel_case(name)
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            [
                (!namespace.cache_items.is_empty()).then(|| format!(
                    r#"{base_indent}  cache = {{
{}
{base_indent}  }};"#,
                    namespace
                        .cache_items
                        .iter()
                        .map(|(name, item)| {
                            let payload = type_item_to_code(&item.payload, level + 2);
                            format!(
                                "{base_indent}    {name}: this.getCacheItem{key}{api}>(\"{name}\"{options}),",
                                key = match &item.key {
                                    None => format!("NoKey<{payload}, "),
                                    Some(key) => {
                                        let key = type_item_to_code(key, level + 2);
                                        format!("<{key}, {payload}, ")
                                    }
                                },
                                api = ALL_CACHE_OPERATIONS
                                    .iter()
                                    .map(|x| match item.expose.contains(x) {
                                        true => "true",
                                        false => "false",
                                    })
                                    .collect::<Vec<_>>()
                                    .join(", "),
                                    options = {
                                        let content = [
                                            item.ttl.as_ref().map(|x| format!(
                                                "{base_indent}      ttl: {},",value_to_code(x))),
                                            item.extend_on_get.as_ref().map(|x| format!(
                                                    "{base_indent}      extendOnGet: {},",value_to_code(x))),
                                        ]
                                        .into_iter()
                                        .flatten()
                                        .collect::<Vec<_>>()
                                        .join("\n");
                                        match content.is_empty() {
                                            true => "".to_string(),
                                            false => format!(
                                                r#", {{
{content}
{base_indent}    }}"#
                                            )
                                        }
                                    }
                            )
                        })
                        .collect::<Vec<_>>()
                        .join("\n")
                )),
                (!namespace.pubsub_items.is_empty()).then(|| format!(
                    r#"{base_indent}  pubsub = {{
{}
{base_indent}  }};"#,
                    namespace
                        .pubsub_items
                        .iter()
                        .map(|(name, item)| {
                            let payload = type_item_to_code(&item.payload, level + 1);
                            format!(
                                "{base_indent}    {name}: this.getPubsubItem{key}{api}>(\"{name}\"),",
                                key = match &item.key {
                                    None => format!("NoKey<{payload}, "),
                                    Some(key) => {
                                        let key = type_item_to_code(key, level + 1);
                                        format!("<{key}, {payload}, ")
                                    }
                                },
                                api = ALL_PUBSUB_OPERATIONS
                                    .iter()
                                    .map(|x| match item.expose.contains(x) {
                                        true => "true",
                                        false => "false",
                                    })
                                    .collect::<Vec<_>>()
                                    .join(", ")
                            )
                        })
                        .collect::<Vec<_>>()
                        .join("\n")
                )),
                (!namespace.task_items.is_empty()).then(|| format!(
                    r#"{base_indent}  task = {{
{}
{base_indent}  }};"#,
                    namespace
                        .task_items
                        .iter()
                        .map(|(name, item)| {
                            let payload = type_item_to_code(&item.payload, level + 1);
                            format!(
                                "{base_indent}    {name}: this.getTaskItem{key}{api}>(\"{name}\"{options}),",
                                key = match &item.key {
                                    None => format!("NoKey<{payload}, "),
                                    Some(key) => {
                                        let key = type_item_to_code(key, level + 1);
                                        format!("<{key}, {payload}, ")
                                    }
                                },
                                api = ALL_TASK_OPERATIONS
                                    .iter()
                                    .map(|x| match item.expose.contains(x) {
                                        true => "true",
                                        false => "false",
                                    })
                                    .collect::<Vec<_>>()
                                    .join(", "),
                                    options = {
                                        let content = [
                                            item.queue_type.as_ref().map(|x| format!(
                                                "{base_indent}      queueType: {},",value_to_code(x))),
                                        ]
                                        .into_iter()
                                        .flatten()
                                        .collect::<Vec<_>>()
                                        .join("\n");
                                        match content.is_empty() {
                                            true => "".to_string(),
                                            false => format!(
                                                r#", {{
{content}
{base_indent}    }}"#
                                            )
                                        }
                                    }
                            )
                        })
                        .collect::<Vec<_>>()
                        .join("\n")
                )),
            ]
            .into_iter()
            .flatten()
            .collect::<Vec<_>>()
            .join("\n")
        ]
        .into_iter()
        .filter(|x| !x.is_empty())
        .collect::<Vec<_>>()
        .join("\n\n"),
    ));
    result
}

pub fn codegen(export_schema: &ExportSchema) -> String {
    format!(
        r#"// deno-fmt-ignore-file
// deno-lint-ignore-file
/* eslint-disable */
import {{ MemorixBase, getEnvVariable }} from "@memorix/client-redis";

{}"#,
        namespace_to_code(
            &export_schema.global_namespace,
            vec![],
            &export_schema.engine,
        )
    )
    .to_string()
}
