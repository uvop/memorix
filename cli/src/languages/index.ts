import { assertUnreachable } from "src/core/utilities";
import { codegenTs } from "./ts/ts";
import { codegenPython } from "./python/python";

export enum Languages {
  typescript = "typescript",
  python = "python",
}

type CodegenByLanguageFn = (schema: string, language: Languages) => string;

export const codegenByLanguage: CodegenByLanguageFn = (scope, language) => {
  switch (language) {
    case Languages.typescript:
      return codegenTs(scope);
    case Languages.python:
      return codegenPython(scope);
    default:
      assertUnreachable(language);
      return "";
  }
};
