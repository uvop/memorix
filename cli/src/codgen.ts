import path from "path";
import fs from "fs";
import { codegenByLanguage } from "./languages";
import { Block, BlockTypes, getBlocks } from "./core/block";

const getSchemaBlocksNoConfig = async (
  schemaFilePath: string,
  dirname?: string
): Promise<Block[]> => {
  const schemaPath =
    dirname !== undefined
      ? path.resolve(dirname, schemaFilePath)
      : path.resolve(schemaFilePath);
  const schemaFolder = path.dirname(schemaPath);
  const schema = await (await fs.promises.readFile(schemaPath)).toString();
  const blocks = getBlocks(schema);
  const blockConfig = blocks.find((x) => x.type === BlockTypes.config);
  if (
    !blockConfig ||
    blockConfig.type !== BlockTypes.config ||
    !blockConfig.extends
  ) {
    return blocks;
  }

  const blocksToAdd = await Promise.all(
    (Array.isArray(blockConfig.extends)
      ? blockConfig.extends
      : [blockConfig.extends]
    ).map(async (schemaExtendFilePath) =>
      getSchemaBlocksNoConfig(schemaFolder, schemaExtendFilePath)
    )
  );

  return [
    ...blocksToAdd.flat(),
    ...blocks.filter((x) => x.type !== BlockTypes.config),
  ];
};

export const codegen = async ({
  schemaFilePath,
}: {
  schemaFilePath: string;
}) => {
  const schemaPath = path.resolve(schemaFilePath);
  const schemaDirname = path.dirname(schemaPath);
  const schema = await (await fs.promises.readFile(schemaPath)).toString();
  const blockConfig = getBlocks(schema).find(
    (x) => x.type === BlockTypes.config
  );
  if (!blockConfig || blockConfig.type !== BlockTypes.config) {
    throw new Error(
      `Must set a Config block to top level schema "${schemaPath}"`
    );
  }
  const otherBlocks = await getSchemaBlocksNoConfig(schemaFilePath);
  const blocks = [blockConfig, ...otherBlocks];

  await Promise.all(
    Array.isArray(blockConfig.output)
      ? blockConfig.output
      : [blockConfig.output].map(async ({ language, file }) => {
          const filePath = path.resolve(schemaDirname, file);
          const code = codegenByLanguage(blocks, language);
          await fs.promises.writeFile(filePath, code);
        })
  );
};
