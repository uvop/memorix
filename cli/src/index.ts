#!/usr/bin/env node
import yargs from "yargs/yargs";
import { hideBin } from "yargs/helpers";
import { Languages } from "./languages";
import { codegen } from "./codgen";

yargs(hideBin(process.argv))
  .command(
    "codegen <schemaFilePath> <...files>",
    "Codegen memorix schema to code",
    (b) => {
      return b
        .positional("schemaFilePath", {
          describe: "Memorix schema file",
        })
        .positional("files", {
          describe: "Pairs of language to codegen and destination file",
        });
    },
    (argv) => {
      const { schemaFilePath, files } = argv as {
        schemaFilePath: string;
        files: string[];
      };
      const parsedFiles: Parameters<typeof codegen>[0]["files"] = [];
      for (let index = 0; index < files.length; index += 2) {
        const languageStr = files[index];
        const dist = files[index + 1];
        const language = Languages[languageStr];
        if (language === undefined) {
          throw new Error(
            `Unknown language, got "${languageStr}", expected one of "${Object.keys(
              Languages
            ).join(", ")}"`
          );
        }
        if (dist === undefined) {
          throw new Error(`Must pass path for "${language}" codegen`);
        }
        parsedFiles.push({
          dist,
          language,
        });
      }
      if (parsedFiles.length === 0) {
        throw new Error(`Didn't get any languages to codegen`);
      }
      codegen({
        schemaFilePath,
        files: parsedFiles,
      });
    }
  )
  .option("verbose", {
    alias: "v",
    type: "boolean",
    description: "Run with verbose logging",
  })
  .demandCommand(1)
  .parse();
