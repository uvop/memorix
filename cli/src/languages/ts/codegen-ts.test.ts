import { Languages, codegenByLanguage } from "src/languages";

const codegenTs = (schema: string) =>
  codegenByLanguage(schema, Languages.typescript).trim();

describe("ts codegen", () => {
  describe("model", () => {
    it("can generate from model", () => {
      expect(
        codegenTs(
          `
            Model User {
                id: int
                name: string
                age: float?
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate from 2 models", () => {
      expect(
        codegenTs(
          `
            Model User1 {
                id: int
            }
            Model User2 {
                name: string
                age: int?
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model within model", () => {
      expect(
        codegenTs(
          `
            Model User {
                id: int
                papa: {
                    name: string?
                }
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model within model within a model", () => {
      expect(
        codegenTs(
          `
            Model User {
                id: int
                papa: {
                    name: string?
                    mama: {
                        sick: boolean
                    }?
                    age: int
                }
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model with array", () => {
      expect(
        codegenTs(
          `
            Model User {
                id: int
                names: [string]
                children: [{
                  id: int
                  name: string?
                }?]?
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("cache", () => {
    it("can generate with inline types", () => {
      expect(
        codegenTs(
          `
            Cache {
              user {
                key: int
                payload: string
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with inline object type", () => {
      expect(
        codegenTs(
          `
            Cache {
              user {
                key: int
                payload: {
                  name: string
                  age: int?
                }?
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with no key", () => {
      expect(
        codegenTs(
          `
            Cache {
              user {
                payload: {
                  name: string
                  age: int?
                }?
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate also with model", () => {
      expect(
        codegenTs(
          `
          Cache {
            adminId {
                payload: string?
            }
            user {
                key: string
                payload: User
            }
          }
        
          Model User {
            name: string
            age: int?
          }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("pubsub", () => {
    it("can generate with inline types", () => {
      expect(
        codegenTs(
          `
            PubSub {
              message {
                key: int
                payload: string
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("task", () => {
    it("can generate with inline types", () => {
      expect(
        codegenTs(
          `
            Task {
              doIt {
                key: int
                payload: string
                returns: boolean
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("enum", () => {
    it("can generate", () => {
      expect(
        codegenTs(
          `
            Enum Animals {
              dog
              cat
              person
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
});
