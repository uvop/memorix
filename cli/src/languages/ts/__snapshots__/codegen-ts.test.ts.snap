// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ts codegen cache can generate also with model 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export type User = {
  name: string;
  age?: number;
};

export class MemorixApi extends MemorixClientApi {
  cache = {
    adminId: this.getCacheItemNoKey<string | undefined>(\\"adminId\\"),
    user: this.getCacheItem<string, User>(\\"user\\"),
  };
}"
`;

exports[`ts codegen cache can generate with inline object type 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi {
  cache = {
    user: this.getCacheItem<number, {
      name: string;
      age?: number;
    } | undefined>(\\"user\\"),
  };
}"
`;

exports[`ts codegen cache can generate with inline types 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi {
  cache = {
    user: this.getCacheItem<number, string>(\\"user\\"),
  };
}"
`;

exports[`ts codegen cache can generate with no key 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi {
  cache = {
    user: this.getCacheItemNoKey<{
      name: string;
      age?: number;
    } | undefined>(\\"user\\"),
  };
}"
`;

exports[`ts codegen cache can have options 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi {
  cache = {
    user: this.getCacheItem<number, string>(\\"user\\", {expire:{value:5}}),
  };
}"
`;

exports[`ts codegen config can be null 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi.fromConfig({defaultOptions:{cache:{}}}) {

}"
`;

exports[`ts codegen config can generate 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi.fromConfig({defaultOptions:{cache:{expire:{value:5}},task:{takeNewest:true}}}) {

}"
`;

exports[`ts codegen enum can generate 1`] = `
"export enum Animals {
  dog = \\"dog\\",
  cat = \\"cat\\",
  person = \\"person\\",
}"
`;

exports[`ts codegen model can generate from 2 models 1`] = `
"export type User1 = {
  id: number;
};

export type User2 = {
  name: string;
  age?: number;
};"
`;

exports[`ts codegen model can generate from model 1`] = `
"export type User = {
  id: number;
  name: string;
  age?: number;
};"
`;

exports[`ts codegen model can generate from model with array 1`] = `
"export type User = {
  id: number;
  names: Array<string>;
  children?: Array<{
    id: number;
    name?: string;
  } | undefined>;
};"
`;

exports[`ts codegen model can generate from model within model 1`] = `
"export type User = {
  id: number;
  papa: {
    name?: string;
  };
};"
`;

exports[`ts codegen model can generate from model within model within a model 1`] = `
"export type User = {
  id: number;
  papa: {
    name?: string;
    mama?: {
      sick: boolean;
    };
    age: number;
  };
};"
`;

exports[`ts codegen pubsub can generate with inline types 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi {
  pubsub = {
    message: this.getPubsubItem<number, string>(\\"message\\"),
  };
}"
`;

exports[`ts codegen task can generate with inline types 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi {
  task = {
    doIt: this.getTaskItem<number, string, boolean>(\\"doIt\\", true),
  };
}"
`;

exports[`ts codegen task can have options 1`] = `
"import { MemorixClientApi } from \\"@memorix/client-redis\\";

export class MemorixApi extends MemorixClientApi {
  task = {
    doIt: this.getTaskItem<number, string, boolean>(\\"doIt\\", true, {takeNewest:true}),
  };
}"
`;
